includes/Modules/AdSense.php

```diff
@@ -93,88 +93,26 @@ function () {
 		add_action(
 			'template_redirect',
 			function() {
-				// Bail early if we are checking for the tag presence from the back end.
-				if ( $this->context->input()->filter( INPUT_GET, 'tagverify', FILTER_VALIDATE_BOOLEAN ) ) {
-					return;
-				}
-
-				if ( $this->is_tag_blocked() ) {
-					return;
-				}
-
-				$use_snippet = $this->get_data( 'use-snippet' );
-				if ( is_wp_error( $use_snippet ) || ! $use_snippet ) {
-					return;
-				}
-
-				$client_id = $this->get_data( 'client-id' );
-				if ( is_wp_error( $client_id ) || ! $client_id ) {
-					return;
-				}
-
 				// Web Stories support neither <amp-auto-ads> nor the script.
 				// TODO: 'amp_story' support can be phased out in the long term.
 				if ( is_singular( array( 'web-story', 'amp_story' ) ) ) {
 					return;
 				}
 
-				// At this point, we know the tag should be rendered, so let's take care of it
-				// for AMP and non-AMP.
-				if ( $this->context->is_amp() ) {
-					add_action( // For AMP Reader, and AMP Native and Transitional (if `wp_body_open` supported).
-						'wp_body_open',
-						function() use ( $client_id ) {
-							$this->print_amp_auto_ads( $client_id );
-						},
-						-9999
-					);
+				$module_settings = $this->get_settings();
+				$settings        = $module_settings->get();
 
-					add_filter( // For AMP Reader, and AMP Native and Transitional (as fallback).
-						'the_content',
-						function( $content ) use ( $client_id ) {
-							// Only run for the primary application of the `the_content` filter.
-							if ( ! in_the_loop() ) {
-								return $content;
-							}
-							return $this->amp_content_add_auto_ads( $content, $client_id );
-						}
-					);
+				$tag = $this->context->is_amp()
+					? new AMP_Tag( $settings['clientID'], self::MODULE_SLUG )
+					: new Web_Tag( $settings['clientID'], self::MODULE_SLUG );
 
-					add_filter( // Load amp-auto-ads component for AMP Reader.
-						'amp_post_template_data',
-						function( $data ) {
-							return $this->amp_data_load_auto_ads_component( $data );
-						}
-					);
+				if ( ! $tag->is_tag_blocked() ) {
+					$tag->use_guard( new TagVerifyGuard( $this->context->input() ) );
+					$tag->use_guard( new Tag_Guard( $module_settings ) );
```

Comment:
Instead of instantiating the `*_Tag` instances here throughout the three module classes, please follow the ACs here, implementing a `Module_With_Tags` interface that would return the complete instance and allows us to remove all the `template_redirect` hooks entirely, replaced with one central piece of invocation. This is also specified in more details in the IB.

While we diverged from the original ACs/IB a bit, I don't think we discussed changing that part, and I think that should be kept as requirement as it simplifies the module-specific implementation. A module would only need to declare itself as `Module_With_Tags` and could then implement the necessary methods.